Excerpts: extract content from inside lists

jeherve opened the pull request:
In #53604, we made some changes so when automatically generating an excerpt from post content, we would better consider content inside inner blocks. r51382 added support for two main wrappers used in Core to house inner blocks: columns and groups.

This commit adds support for another commonly used wrapper in core: lists. This way, content from lists would be pulled in excerpts when possible.

Trac ticket: https://core.trac.wordpress.org/ticket/63909

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

jeherve committed:
Excerpts: extract content from inside lists

In #53604, we made some changes so when automatically generating an excerpt from post content, we would better consider content inside inner blocks. r51382 added support for two main wrappers used in Core to house inner blocks: columns and groups.

This commit adds support for another commonly used wrapper in core: lists. This way, content from lists would be pulled in excerpts when possible.

Trac ticket: https://core.trac.wordpress.org/ticket/63909

mukeshpanchal27 reviewed (approved):
Thanks @jeherve for the PR. Looks good to me

audrasjb committed:
Merge branch 'trunk' into add/wp-trim-excerpt-lists

huzaifaalmesbah reviewed (approved):
Looks good.

https://github.com/user-attachments/assets/906375b8-5fce-4363-a837-fea41ca56558

t-hamano commented on src/wp-includes/blocks.php:
```diff
@@ -2195,7 +2169,7 @@ function excerpt_remove_blocks( $content ) {
 		'core/heading',
 		'core/html',
-		'core/list',
+		'core/list-item',
```
```suggestion
		'core/list',
		'core/list-item',
```

The List block previously did not have inner blocks. Therefore, if we remove `core/list` here, those legacy List blocks will no longer appear in excerpts.

t-hamano commented on tests/phpunit/tests/post/getTheExcerpt.php:
It would also be great to have a test that verifies that legacy List blocks are output in the excerpt.

For example, like this.

```php
public function test_legacy_list_block_excerpt() {
	$content = '<!-- wp:quote -->
<blockquote class="wp-block-quote"><!-- wp:list -->
<ul><li>Quoted list item.</li></ul>
<!-- /wp:list --></blockquote>
<!-- /wp:quote -->';

	$post = self::factory()->post->create_and_get(
		array(
			'post_content' => $content,
			'post_excerpt' => '',
		)
	);

	$this->assertSame( 'Quoted list item. Quoted paragraph.', get_the_excerpt( $post ) );
}
```

jeherve replied:
Sure ting, can do. Done in bd01d0f271156b978dcea1a72e324b4bcbaab6b6

jeherve committed:
Update src/wp-includes/blocks.php

Co-authored-by: Aki Hamano <54422211+t-hamano@users.noreply.github.com>

jeherve committed:
Excerpts: cover legacy List blocks in tests

Legacy List blocks keep their items as raw HTML rather than inner
blocks, so they take a different path through excerpt_remove_blocks()
than the current markup does. Nothing pinned that behaviour down,
and it is exactly what breaks if core/list stops being an allowed
inner block.

Add a test for a legacy list on its own and nested inside a quote.
Removing core/list from the allowed inner blocks fails the nested
case, which is the regression we want to hear about.

jeherve committed:
Merge remote-tracking branch 'upstream/trunk' into add/wp-trim-excerpt-lists
